File: ltviews/src/main/java/com/lt/ltviews/lt_recyclerview/MTextRefreshLayout.java
package com.lt.ltviews.lt_recyclerview;

import android.animation.ObjectAnimator;
import android.content.Context;
import android.content.SharedPreferences;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.LinearInterpolator;
import android.widget.ImageView;
import android.widget.TextView;

import com.lt.ltviews.R;

import java.text.SimpleDateFormat;
import java.util.Date;


/**
 * 创    建:  lt  2018/5/23--18:42
 * 作    用:  文字的下拉刷新
 * 注意事项:
 */

public class MTextRefreshLayout extends LtRefreshLayout {

    TextView tv;
    TextView tvDate;
    ImageView iv;
    ObjectAnimator oa;
    String date;
    SimpleDateFormat sdf;

    public MTextRefreshLayout(Context context) {
        this(context, null);
    }

    public MTextRefreshLayout(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public MTextRefreshLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        sdf = new SimpleDateFormat("M-d H:m");
        date = getDate();
    }

    @Override
    protected void onStatus(int status) {
        switch (status) {
            case REFRESH_DOWN:
                //下拉中
                tv.setText("下拉可以刷新");
                iv.setVisibility(View.VISIBLE);
                iv.setImageResource(R.drawable.lt_arrow);
                //设置动画时间
                oa = ObjectAnimator.ofFloat(iv, "rotation", iv.getRotation(), 0)
                        .setDuration(animationTime);
                oa.setRepeatCount(0);//设置动画执行的次数
                oa.start();//开始动画
                break;
            case REFRESH_RELEASE:
                //松开刷新
                tv.setText("释放立即刷新");
                iv.setVisibility(View.VISIBLE);
                iv.setImageResource(R.drawable.lt_arrow);
                oa = ObjectAnimator.ofFloat(iv, "rotation", iv.getRotation(), 180)
                        .setDuration(animationTime);//设置动画时间
                oa.setRepeatCount(0);//设置动画执行的次数
                oa.start();//开始动画
                break;
            case REFRESHING:
                //刷新中
                tv.setText("正在刷新...");
                iv.setVisibility(View.VISIBLE);
                iv.setImageResource(R.drawable.lt_loading);
                oa = ObjectAnimator.ofFloat(iv, "rotation", 0, 360)
                        .setDuration(animationTime << 2);//设置动画时间
                oa.setInterpolator(new LinearInterpolator());
                oa.setRepeatCount(ObjectAnimator.INFINITE);//设置动画执行的次数,这个是无限
                oa.start();//开始动画
                break;
            case REFRESH_FINISH:
                //刷新完成
                date = setDate();
                tv.setText("刷新完成");
                tvDate.setText("上次更新: " + date);
                iv.clearAnimation();
                oa.cancel();
                iv.setVisibility(View.INVISIBLE);
                iv.setImageResource(R.drawable.lt_arrow);
                break;
            default:
                break;
        }
    }

    @Override
    protected void onProgress(float y) {

    }

    @Override
    protected View getRefreshView() {
        View view = View.inflate(getContext(), R.layout.lt_refresh_view, null);
        tv = (TextView) view.findViewById(R.id.tv_lt_refresh);
        tvDate = (TextView) view.findViewById(R.id.tv_lt_date);
        iv = (ImageView) view.findViewById(R.id.iv_lt_refresh);
        tvDate.setText("上次更新: " + date);
        return view;
    }

    String getDate() {
        SharedPreferences preference = getContext().getSharedPreferences("lt_rv",
                Context.MODE_PRIVATE);
        return preference.getString("lt_date", "");
    }

    String setDate() {
        String value = sdf.format(new Date());
        SharedPreferences preference = getContext().getSharedPreferences("lt_rv",
                Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = preference.edit();
        editor.putString("lt_date", value);
        editor.commit();
        return value;
    }
}

File: ltviewsx/src/main/java/com/lt/ltviewsx/lt_recyclerview/MTextRefreshLayout.kt
package com.lt.ltviewsx.lt_recyclerview

import android.animation.ObjectAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.animation.LinearInterpolator
import android.widget.ImageView
import android.widget.TextView
import com.lt.ltviewsx.R
import java.text.SimpleDateFormat
import java.util.*

/**
 * 创    建:  lt  2018/5/23--18:42
 * 作    用:  文字的下拉刷新
 * 注意事项:
 */
class MTextRefreshLayout @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0)
    : LtRefreshLayout(context, attrs, defStyleAttr) {
    private var tv: TextView? = null
    private var tvDate: TextView? = null
    private var iv: ImageView? = null
    private var oa: ObjectAnimator? = null
    private var date: String = justGetDate()
    private var sdf: SimpleDateFormat = SimpleDateFormat("M-d H:m")

    @SuppressLint("ObjectAnimatorBinding")
    override fun onState(state: RefreshStates) {
        when (state) {
            RefreshStates.STATE_REFRESH_DOWN -> {
                //下拉中
                tv?.setText(R.string.down_refresh)
                iv?.visibility = View.VISIBLE
                iv?.setImageResource(R.drawable.lt_arrow)
                //设置动画时间
                oa = ObjectAnimator.ofFloat(iv, "rotation", iv?.rotation ?: 0F, 0f)
                        .setDuration(animationTime.toLong())
                oa?.repeatCount = 0 //设置动画执行的次数
                oa?.start() //开始动画
            }
            RefreshStates.STATE_REFRESH_RELEASE -> {
                //松开刷新
                tv?.setText(R.string.release_refresh_now)
                iv?.visibility = View.VISIBLE
                iv?.setImageResource(R.drawable.lt_arrow)
                oa = ObjectAnimator.ofFloat(iv, "rotation", iv?.rotation ?: 0F, 180f)
                        .setDuration(animationTime.toLong()) //设置动画时间
                oa?.setRepeatCount(0) //设置动画执行的次数
                oa?.start() //开始动画
            }
            RefreshStates.STATE_REFRESHING -> {
                //刷新中
                tv?.setText(R.string.refresh)
                iv?.visibility = View.VISIBLE
                iv?.setImageResource(R.drawable.lt_loading)
                oa = ObjectAnimator.ofFloat(iv, "rotation", 0f, 360f)
                        .setDuration((animationTime shl 2.toLong().toInt()).toLong()) //设置动画时间
                oa?.setInterpolator(LinearInterpolator())
                oa?.setRepeatCount(ObjectAnimator.INFINITE) //设置动画执行的次数,这个是无限
                oa?.start() //开始动画
            }
            RefreshStates.STATE_REFRESH_FINISH -> {
                //刷新完成
                date = saveAndGetDate()
                tv?.setText(R.string.refresh_finish)
                tvDate?.text = context.getString(R.string.last_update) + date
                iv?.clearAnimation()
                oa?.cancel()
                iv?.visibility = View.INVISIBLE
                iv?.setImageResource(R.drawable.lt_arrow)
            }
            RefreshStates.STATE_BACK -> {
                //刷新头回到重新隐藏了回去,此时重置箭头的动画
                oa = ObjectAnimator.ofFloat(iv, "rotation", iv?.rotation ?: 0F, 0f)
                        .setDuration(animationTime.toLong())
                oa?.repeatCount = 0 //设置动画执行的次数
                oa?.start() //开始动画
            }
        }
    }

    override fun onProgress(y: Float) {}

    override fun createRefreshView(): View {
        val view = View.inflate(context, R.layout.lt_refresh_view, null)
        tv = view.findViewById<View>(R.id.tv_lt_refresh) as TextView
        tvDate = view.findViewById<View>(R.id.tv_lt_date) as TextView
        iv = view.findViewById<View>(R.id.iv_lt_refresh) as ImageView
        tvDate?.text = context.getString(R.string.last_update) + date
        return view
    }

    private fun justGetDate(): String {
        val preference = context.getSharedPreferences("lt_rv",
                Context.MODE_PRIVATE)
        return preference.getString("lt_date", "")
    }

    private fun saveAndGetDate(): String {
        val value = sdf.format(Date())
        val preference = context.getSharedPreferences("lt_rv",
                Context.MODE_PRIVATE)
        val editor = preference.edit()
        editor.putString("lt_date", value)
        editor.apply()
        return value
    }
}